compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -275,59 +275,84 @@ fn emit_orphan_check_error<'tcx>(
             err.note("define and implement a trait or new type instead");
             err.emit()
         }
-        traits::OrphanCheckErr::UncoveredTy(param_ty, local_type) => {
-            let mut sp = sp;
-            for param in generics.params {
-                if param.name.ident().to_string() == param_ty.to_string() {
-                    sp = param.span;
+        traits::OrphanCheckErr::UncoveredTy(param_ty, idx, local_type) => {
+            let sp;
+            let mut param_ty_in_fundamental_ty = false;
+
+            match idx {
+                0 => {
+                    sp = self_ty_span;
+                    match *self_ty.kind() {
+                        ty::Ref(..) => {
+                            param_ty_in_fundamental_ty = true;
+                        }
+                        ty::Adt(def, _) => {
+                            param_ty_in_fundamental_ty = def.is_fundamental();
+                        }
+                        _ => (),
+                    }
```

Comment:
why are you only doing that if the uncovered ty is in the self type?

you can also have `impl<T> Trait<Box<T>> for ()` which would also benefit from this message